### PR TITLE
NMS-12739: Don't allow to import a requisition with same service names.

### PR DESCRIPTION
**Pull request by jadyhou**
https://issues.opennms.org/browse/NMS-12739

### All Contributors

* [ ] Have you read and followed our [Contribution Guidelines](https://github.com/OpenNMS/opennms/blob/develop/CONTRIBUTING.md)?
* [ ] Have you [made an issue in the OpenNMS issue tracker](https://issues.opennms.org)?<br>If so, you should:
  1. update the title of this PR to be of the format: `${JIRA-ISSUE-NUMBER}: subject of pull request`
  2. update the JIRA link at the bottom of this comment to refer to the real issue number
  3. prefix your commit messages with the issue number, if possible
* [ ] Have you made a comment in that issue which points back to this PR?
* [ ] Have you updated the JIRA link at the bottom of this comment to link to your issue?
* [ ] If this is a new or updated feature, is there documentation for the new behavior?
* [ ] If this is new code, are there unit and/or integration tests?
* [ ] If this PR targets a `foundation-*` branch, does it avoid changing files in `$OPENNMS_HOME/etc/`?

### Pull Request Process

One or more reviewers should be assigned to each PR.

If you know that a particular person is subject matter expert in the area your PR affects, feel free to assign one or more reviewers when you create this PR, otherwise reviewers will be assigned for you.

If you have made additions or changes to the documentation, or if you _need_ documentation for these code changes, please make sure a technical writer has looked it over.

Once the reviewer(s) accept the PR and the branch passes continuous integration, the PR is eligible for merge.

At that time, if you have commit access (are an OpenNMS Group employee or a member of the Order of the Green Polo) you are welcome to merge the PR.
Otherwise, a reviewer can merge it for you.

Thanks for taking time to contribute!

### External References

* JIRA (Issue Tracker): http://issues.opennms.org/browse/${JIRA-ISSUE-NUMBER}

### CODE REVIEW

File: opennms-provision/opennms-provision-persistence/src/main/java/org/opennms/netmgt/provision/persist/requisition/RequisitionInterface.java [hunk 1]
```diff
@@ -398,8 +400,12 @@ public void validate() throws ValidationException {
             throw new ValidationException("Requisition interface 'ip-addr' is a required attribute!");
         }
         if (m_monitoredServices != null) {
+            Set<String> serviceNameSet = new HashSet<>();
             for (final RequisitionMonitoredService svc : m_monitoredServices) {
                 svc.validate();
+                if(!serviceNameSet.add(svc.getServiceName())){
+                    throw new ValidationException("Duplicate service names are not allowed.");
```
Comment (mattixtech) [inline opennms-provision/opennms-provision-persistence/src/main/java/org/opennms/netmgt/provision/persist/requisition/RequisitionInterface.java:407]: We may want to indicate what the duplicate service name we failed with was.

File: opennms-provision/opennms-provision-persistence/src/main/java/org/opennms/netmgt/provision/persist/requisition/RequisitionInterface.java [hunk 2]
```diff
@@ -398,8 +400,12 @@ public void validate() throws ValidationException {
             throw new ValidationException("Requisition interface 'ip-addr' is a required attribute!");
         }
         if (m_monitoredServices != null) {
+            Set<String> serviceNameSet = new HashSet<>();
             for (final RequisitionMonitoredService svc : m_monitoredServices) {
                 svc.validate();
+                if(!serviceNameSet.add(svc.getServiceName())){
```
Comment (mattixtech) [inline opennms-provision/opennms-provision-persistence/src/main/java/org/opennms/netmgt/provision/persist/requisition/RequisitionInterface.java:406]:
```suggestion
                if (!serviceNameSet.add(svc.getServiceName())) {
```

**Review by mattixtech — CHANGES_REQUESTED** [review]
It might be worth adding a unit test to cover this case. Not sure if there is already convenient tests for this that you could simply add to.

**Comment by mattixtech** [comment]
Make sure to update the PR description to include a valid link to the Jira rather than the placeholder link.

**Comment by cgorantla** [comment]
Probably we can fix [this](https://github.com/OpenNMS/opennms/blob/d562afb4fa4ca8362c6fcd65caf24cc72d30ee86/opennms-provision/opennms-provision-persistence/src/main/java/org/opennms/netmgt/provision/persist/requisition/RequisitionInterface.java#L67)  which allows us unique elements.  We don't need to even fail. Just ignore.

You can add a test for this specific case [here](https://github.com/OpenNMS/opennms/blob/d562afb4fa4ca8362c6fcd65caf24cc72d30ee86/opennms-webapp-rest/src/test/java/org/opennms/web/rest/v1/RequisitionRestServiceIT.java#L72)

**Comment by jadyhou** [comment]
> Probably we can fix [this](https://github.com/OpenNMS/opennms/blob/d562afb4fa4ca8362c6fcd65caf24cc72d30ee86/opennms-provision/opennms-provision-persistence/src/main/java/org/opennms/netmgt/provision/persist/requisition/RequisitionInterface.java#L67) which allows us unique elements. We don't need to even fail. Just ignore.
> 
> You can add a test for this specific case [here](https://github.com/OpenNMS/opennms/blob/d562afb4fa4ca8362c6fcd65caf24cc72d30ee86/opennms-webapp-rest/src/test/java/org/opennms/web/rest/v1/RequisitionRestServiceIT.java#L72)

That would be another option. But the thing that could be duplicated is here(not a list):

    @XmlAttribute(name="service-name", required=true)
    protected String m_serviceName;

but not here actually:
    @XmlElement(name="category")
    protected List<RequisitionCategory> m_categories = new ArrayList<RequisitionCategory>();

**Comment by RangerRick** [comment]
Might not be a bad idea to also make the XSD require uniqueness, so it fails at the XML level: https://support.liquid-technologies.com/kb/a79/creating-a-unique-constraint-with-an-xsd.aspx

**Comment by jadyhou** [comment]
> Might not be a bad idea to also make the XSD require uniqueness, so it fails at the XML level: https://support.liquid-technologies.com/kb/a79/creating-a-unique-constraint-with-an-xsd.aspx

Thanks, will take a look at the XSD.

**Comment by cgorantla** [comment]
> Probably we can fix [this](https://github.com/OpenNMS/opennms/blob/d562afb4fa4ca8362c6fcd65caf24cc72d30ee86/opennms-provision/opennms-provision-persistence/src/main/java/org/opennms/netmgt/provision/persist/requisition/RequisitionInterface.java#L67) which allows us unique elements. We don't need to even fail. Just ignore.
> 
> You can add a test for this specific case [here](https://github.com/OpenNMS/opennms/blob/d562afb4fa4ca8362c6fcd65caf24cc72d30ee86/opennms-webapp-rest/src/test/java/org/opennms/web/rest/v1/RequisitionRestServiceIT.java#L72)

I'm okay with returning error for duplicate service. In that context, changes look good to me. 
Just need to add test.

File: opennms-provision/opennms-provision-persistence/src/main/java/org/opennms/netmgt/provision/persist/requisition/RequisitionInterface.java (hunk 1 above)
Comment (jadyhou) [reply opennms-provision/opennms-provision-persistence/src/main/java/org/opennms/netmgt/provision/persist/requisition/RequisitionInterface.java]: True. That'll be a better user experience. Addressed.

File: opennms-provision/opennms-provision-persistence/src/main/java/org/opennms/netmgt/provision/persist/requisition/RequisitionInterface.java (hunk 2 above)
Comment (jadyhou) [reply opennms-provision/opennms-provision-persistence/src/main/java/org/opennms/netmgt/provision/persist/requisition/RequisitionInterface.java]: Updated.

**Comment by jadyhou** [comment]
> > Probably we can fix [this](https://github.com/OpenNMS/opennms/blob/d562afb4fa4ca8362c6fcd65caf24cc72d30ee86/opennms-provision/opennms-provision-persistence/src/main/java/org/opennms/netmgt/provision/persist/requisition/RequisitionInterface.java#L67) which allows us unique elements. We don't need to even fail. Just ignore.
> > You can add a test for this specific case [here](https://github.com/OpenNMS/opennms/blob/d562afb4fa4ca8362c6fcd65caf24cc72d30ee86/opennms-webapp-rest/src/test/java/org/opennms/web/rest/v1/RequisitionRestServiceIT.java#L72)
> 
> I'm okay with returning error for duplicate service. In that context, changes look good to me.
> Just need to add test.

Test case added.

**Review by cgorantla — APPROVED** [review]
LGTM

**Review by mattixtech — APPROVED** [review]

**Comment by jadyhou** [comment]
> Might not be a bad idea to also make the XSD require uniqueness, so it fails at the XML level: https://support.liquid-technologies.com/kb/a79/creating-a-unique-constraint-with-an-xs

I'm merging this for now because I haven't made this unique happen in xml level. And fail in xml level end up the REST call return 500 server error but not 400 bad request which is not right. This could be another Jira.

**Pull request merged**